internal/services/notification/webhooks.go

```diff
@@ -0,0 +1,217 @@
+
+	webhookVersion = 1
+)
+
+type RunWebhook struct {
+	// version of the webhook struct data
+	Version uint64 `json:"version"`
+
+	// info of the project
+	ProjectInfo ProjectInfo `json:"project_info"`
+
+	// current run status
+	Run *Run `json:"run"`
+
+	// the trigger of the event
+	// (i.e. the value run_phase_changed indicate a run phase changed event)
+	EventType string `json:"event_type"`
+}
+
+type ProjectInfo struct {
+	// id of the project
+	ProjectID string `json:"project_id"`
+}
+
+type Run struct {
+	ID          string              `json:"id"`
+	Annotations map[string]string   `json:"annotations"`
+	Name        string              `json:"name"`
+	Counter     uint64              `json:"counter"`
+	Phase       string              `json:"phase"`
+	Result      string              `json:"result"`
+	SetupErrors []string            `json:"setup_errors"`
+	Tasks       map[string]*RunTask `json:"tasks"`
+	EnqueueTime *time.Time          `json:"enqueue_time"`
+	StartTime   *time.Time          `json:"start_time"`
+	EndTime     *time.Time          `json:"end_time"`
+}
+
+type RunTask struct {
+	ID              string                    `json:"id"`
+	Name            string                    `json:"name"`
+	Level           int                       `json:"level"`
+	Skip            bool                      `json:"skip"`
+	Depends         map[string]*RunTaskDepend `json:"depends"`
+	Status          string                    `json:"status"`
+	Timedout        bool                      `json:"timedout"`
+	WaitingApproval bool                      `json:"waiting_approval"`
+	Approved        bool                      `json:"approved"`
+	SetupStep       RunTaskStep               `json:"setup_step"`
+	Steps           []*RunTaskStep            `json:"steps"`
+	StartTime       *time.Time                `json:"start_time"`
+	EndTime         *time.Time                `json:"end_time"`
+}
+
+type RunTaskStep struct {
+	Phase      string     `json:"phase"`
+	ExitStatus *int       `json:"exit_status"`
+	StartTime  *time.Time `json:"start_time"`
+	EndTime    *time.Time `json:"end_time"`
+}
+
+type RunTaskDepend struct {
+	TaskID     string   `json:"task_id"`
+	Conditions []string `json:"conditions"`
+}
+
+func (n *NotificationService) sendWebhooks(ctx context.Context, ev *rstypes.RunEvent) error {
+	data := ev.Data.(*rstypes.RunEventData)
+
+	// ignore user direct runs
+	if data.Annotations[action.AnnotationRunType] == string(common.GroupTypeUser) {
+		return nil
+	}
+
+	webhook := &RunWebhook{
+		Version: webhookVersion,
+		ProjectInfo: ProjectInfo{
+			ProjectID: data.Annotations[action.AnnotationProjectID],
+		},
+		EventType: string(ev.RunEventType),
+		Run:       &Run{},
+	}
+
+	webhook.Run.ID = data.ID
+	webhook.Run.Annotations = make(map[string]string)
+	webhook.Run.Annotations[action.AnnotationRefType] = data.Annotations[action.AnnotationRefType]
+	webhook.Run.Annotations[action.AnnotationRef] = data.Annotations[action.AnnotationRef]
```

Comment:
You did the opposite of the previous comment. You should export ALL annotations as fields.